src/blockMonitor/confirmationObserver.ts

```diff
@@ -0,0 +1,130 @@
+import { StartStopService, ArgumentError, BlockThresholdReachedError, ReorgError } from "../dataEntities";
```

Comment:
As I mentioned before, I think the logic in this class is a bit tricky to understand. Try refactoring it to make it clearer, the main thing I'm thinking about is that the class does too much in a single function. Start by separating the functionality, then you can find a nice way of recombining them afterwards to make it easy for a caller - or maybe leave the separate for more flexibility.

Reply:
Simplified quite a lot in 11331a6e57a354cc70f69f81cbaff4c5fcf0237f, by splitting waitForConfirmations into two separate functions: waitForFirstConfirmation (that optionally throws if the transaction is stuck, but does not care about reorgs), and waitForConfirmations(that optionally throws for Reorg, but assumes that the transaction was already confirmed).

There is still a bit too much CancellablePromise-related boilerplate, but I think it is manageable, and much easier to read now.